File: lib/youtube.ts
const PATTERNS = [
  /(?:youtube\.com|youtube-nocookie\.com)\/watch\?(?:.*&)?v=([A-Za-z0-9_-]{11})/,
  /youtu\.be\/([A-Za-z0-9_-]{11})/,
  /youtube\.com\/shorts\/([A-Za-z0-9_-]{11})/,
  /youtube\.com\/embed\/([A-Za-z0-9_-]{11})/,
];

export function parseYouTubeId(url: string): string | null {
  if (!url?.trim()) return null;
  for (const p of PATTERNS) {
    const m = url.match(p);
    if (m) return m[1];
  }
  return null;
}

/** hqdefault always exists; maxresdefault 404s on many videos. Do not "upgrade" this. */
export function thumbnailUrl(id: string): string {
  return `https://i.ytimg.com/vi/${id}/hqdefault.jpg`;
}

export function embedUrl(id: string): string {
  return `https://www.youtube-nocookie.com/embed/${id}?autoplay=1&rel=0&modestbranding=1`;
}

File: api/_lib/youtube.py
import re

_PATTERNS = [
    re.compile(r"(?:youtube\.com|youtube-nocookie\.com)/watch\?(?:.*&)?v=([A-Za-z0-9_-]{11})"),
    re.compile(r"youtu\.be/([A-Za-z0-9_-]{11})"),
    re.compile(r"youtube\.com/shorts/([A-Za-z0-9_-]{11})"),
    re.compile(r"youtube\.com/embed/([A-Za-z0-9_-]{11})"),
]


def parse_youtube_id(url: str) -> str:
    """Extract the 11-character video ID. Raises ValueError if absent."""
    if not isinstance(url, str) or not url.strip():
        raise ValueError("empty url")
    for pattern in _PATTERNS:
        if match := pattern.search(url):
            return match.group(1)
    raise ValueError(f"not a recognisable YouTube URL: {url!r}")
